misc/codepoint_width: add unicode width detection support

kasper93 opened the pull request.

kasper93 commented:
/cc @guidocella

avih commented:
The function function `codepoint_width` sounds and looks eerily similar to `wcwidth` as specified by POSIX, but it's named differently. It should be documented someplace exactly what it does, or, alternatively, how exactly it differers from POSIX `wcwidth`.

> enerated by GraphemeTableGen from Microsoft Terminal (MIT Licence):
> https://github.com/microsoft/terminal/blob/a7e47b711a2adc7b9e80eddea8168089f7d3b11e/src/tools/GraphemeTableGen/Program.cs
>
> With minor adjustment to use it in C codebase.

If tomorrow someone wants to update the data for Unicode 16.1, it's really unclear from the commit message what are the steps to take to end up with a new data file.

kasper93 commented:
> The function `codepoint_width` sounds and looks eerily similar to `wcwidth` as specified by POSIX, but it's named differently. It should be documented someplace exactly what it does, or, alternatively, how exactly it differs from POSIX `wcwidth`.

I don't know why I should adhere to POSIX here in any shape or form.

> > Generated by GraphemeTableGen from Microsoft Terminal (MIT Licence):
> > https://github.com/microsoft/terminal/blob/a7e47b711a2adc7b9e80eddea8168089f7d3b11e/src/tools/GraphemeTableGen/Program.cs
> > With minor adjustment to use it in C codebase.
> 
> If tomorrow someone wants to update the data for Unicode 16.1, it's really unclear from the commit message what are the steps to take to end up with a new data file.
> 
> Do they need to compile `Program.cs`? Any specific options to use when running this program? what should they change at the resulting file?

Yes, one need to run GraphemeTableGen generator and copy the output to mpv's codebase.

But if you really want how to do it:
1. Get GraphemeTableGen source code
2. Compile GraphemeTableGen source code
3. Run GraphemeTableGen executable
4. Copy GraphemeTableGen generated tables to mpv codebase
5. Remove constexpr and add missing includes if needed

avih commented:
>> The function codepoint_width sounds and looks eerily similar to wcwidth as specified by POSIX, but it's named differently. It should be documented someplace exactly what it does, or, alternatively, how exactly it differs from POSIX wcwidth.

> I don't know why I should adhere to POSIX here in any shape or form.

I didn't say you should. It was one option.

Because they already have a spec, it might be easier to explain what it does by saying how it differs from an existing spec.

Otherwise, you should specify what it does, without referring to POSIX.

If you don't do any of those, then no can tell what it does or how to use it. They'll have to guess. Which is not great.

> Yes, one need to run GraphemeTableGen generator and copy the output to mpv's codebase.
> 
> But if you really want how to do it:
> ...

Not me. Anyone which wants to update to the next Unicode spec once it comes out and you don't have time to handle it yourself.

So the steps should be at the commit message or the file itself.

What if they only have Linux or BSD system?

kasper93 commented:
> If you don't do any of those, then no one can tell what it does or how to use it. They'll have to guess. Which is not great.

Sure, I can add, but it is not meant to be used by anything else, except msg.c, nor it should be ever considered to be used as-is anywhere else.

> What if they only have Linux or BSD system?

This is a simple app, that can be run with Mono anywhere. https://en.wikipedia.org/wiki/Mono_(software)

avih commented:
> This is a simple app, that can be run with Mono anywhere. https://en.wikipedia.org/wiki/Mono_(software)

Thanks for the info.

Maybe it can run in a lot of places, but it can't be compiled with mono on Ubuntu 22.04:
```
$ mono --version
Mono JIT compiler version 6.8.0.105 (Debian 6.8.0.105+dfsg-3.2 Wed Jun 30 05:34:49 UTC 2021)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        Suspend:       hybrid
        GC:            sgen (concurrent by default)


$ mcs ms-wcwidth-gen.cs
ms-wcwidth-gen.cs(5,17): error CS1041: Identifier expected, `uint' is a keyword
ms-wcwidth-gen.cs(43,5): error CS1001: Unexpected symbol `]', expecting identifier
ms-wcwidth-gen.cs(43,7): error CS1001: Unexpected symbol `]', expecting identifier
ms-wcwidth-gen.cs(43,9): error CS1001: Unexpected symbol `]', expecting identifier
ms-wcwidth-gen.cs(46,4): error CS1001: Unexpected symbol `[', expecting identifier
ms-wcwidth-gen.cs(49,32): error CS1001: Unexpected symbol `[', expecting identifier
ms-wcwidth-gen.cs(49,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(50,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(51,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(52,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(53,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(54,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(55,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(56,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(57,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(58,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(59,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(60,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(61,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(62,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(69,32): error CS1001: Unexpected symbol `[', expecting identifier
ms-wcwidth-gen.cs(69,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(70,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(71,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(72,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(73,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(74,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(75,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(76,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(77,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(78,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(79,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(80,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(81,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(82,34): warning CS0658: `value' is invalid attribute target. All attributes in this attribute section will be ignored
ms-wcwidth-gen.cs(89,11): error CS1010: Newline in constant
ms-wcwidth-gen.cs(94,11): error CS1010: Newline in constant
ms-wcwidth-gen.cs(97,1): error CS1518: Attributes must be attached to class, delegate, enum, interface or struct
ms-wcwidth-gen.cs(97,0): error CS1001: Unexpected symbol `}', expecting identifier
Compilation failed: 11 error(s), 28 warnings

$
```

kasper93 commented:
Maybe ancient mono is not enough. This works.

```
sudo apt-get install dotnet-sdk-8.0
dotnet build GraphemeTableGen.csproj
./bin/Debug/net8.0/GraphemeTableGen
```

avih commented:
I'm wondering out loud, instead of requiring the dotnet sdk or visual studio, maybe the instructions should be more like:
- Create an empty file `codepoint_width.c`
- Copy into it from the file https://github.com/microsoft/terminal/blob/main/src/types/CodepointWidthDetector.cpp (maybe a specific version of it): all the `s_stage*[]` arrays, and the function `ucdLookup`
- Do this and that ... (includes, rename the function, whatever, to end up how we want it)

And also mention that it can be generated using https://github.com/microsoft/terminal/blob/main/src/tools/GraphemeTableGen/Program.cs but it requires visual studio/mono/dotnet-sdk/whatever...

avih commented on misc/codepoint_width.c:
```diff
@@ -0,0 +1,606 @@
+
+int codepoint_width(const wchar_t cp)
+{
+    const uint16_t s0 = s_stage0[cp >> 11];
```
On Windows `wchar_t` is 16 bits, which can't hold the codepoints of all the nice Emojies from the recent decade.

And on Linux where `wchar_t` is 32 bit, or everywhere if it's changed it to e.g. `uint32_t`, I'm quite sure that `s_stage0[cp >> 11]` can access out of bounds, because there are 21 bits remaining after shifting 11 of 32, and I doubt the array has 2M values, and if it doesn't then an arbitrary 21 bit index is not guaranteed to be in-bound.

In other words, this likely requires a test to ensure the input is within some range.

kasper93 replied:
> which can't hold the codepoints of all the nice Emojies from the recent decade.

They are "broken" on Windows anyway, because there wchar_t is used ubiquitously for wide character strings. But good point, no need to impose this limitation here.

> I doubt the array has 2M values, and if it doesn't then an arbitrary 21 bit index is not guaranteed to be in-bound.

As you probably know, 0x10FFFF is the maximum for valid code point. This is not external API, we already ensure only valid code points are extracted from utf-8 string.

> In other words, this likely requires a test to ensure the input is within some range.

It does not, it is on the caller.

avih commented:
> Because they already have a spec, it might be easier to explain what it does by saying how it differs from an existing spec.

So here's what it does. It's the same as POSIX `wcwidth`, with the following differences:
- It does get most of the width-0 and width-2 codepoints correctly (but see below where it doesn't), and also most of the width-1 codepoints (with one big exception - ambiguous width, see below).
- it returns 1 where `wcwidth` returns `-1`. These are the C0 range, and DEL, and C1 ranges (1-31, 127, 128-159), and the last 2 codepoints in each plane (U+*FFF[EF]), and in some implementations also values which are reserved for surrogate-halves. Arguably it's no biggie because text shouldn't include those, but it's a difference, and it can't be used to detect non-printable codepoints.
- It doesn't consider the categories `Zl` and `Zp` (line/paragraph separator) as zero-width. Only two codepoints: U+2028 and U+2029. No biggie.
- It doesn't consider the `Mc` Unicode category ("Mark, spacing combining") as zero-width, while some (but not all) other `wcwidth` implementations do consider it zero-width. There are 468 codepoints in the `Mc` category. I don't know if it matters much in practice.
- It doesn't consider some Hangul blocks as zero-width. No biggie.
- It doesn't consider the Emoji skin-tone Modifiers U+1F3FB .. U+1F3FF as zero width (instead, it considers it wide/double-width). This can actually be important, because emojis are kinda popular, and this will calculate an incorrect width for emoji+skintone-modifier combo. For instance the combo U+1F469, U+1F3FB (👩🏻) would be incorrectly calculated as width 4 instead of 2.
- And lastly, as far as I can tell, without programmatically testing it conclusively, it returns 3 where the Unicode `East_Asian_Width` property is `A`(mbiguous), and where a general implementation (which doesn't have anything better to do with ambiguous width) should return 1. Either way, a user of this functions is not expecting 3 as a width return value. The function should decide if it wants to return 1 or 2, and generally it should return 1 in this case. There are 138238 codepoints which have ambigous width, and that's more than 10% of the whole Unicode range. So there's that.

Overall, after fixing the type, adding a range check to ensure there's no out-of-bounds access, making it return 1 instead of 3, and maybe fixing the emoji-modifiers, it's OK to use in mpv IMO.

It's probably slower than it could have been for ascii chars, which are arguably the vast majority of the input it would have to deal with, but that's not really an issue IMO in mpv (many implementations return a value for the ascii range without using search/lookup).

You're welcome.

kasper93 commented:
> Overall, after fixing the type

Fixed already.

> adding a range check to ensure there's no out-of-bounds access

This is on the caller and the bstr_decode_utf8(), by design.

> making it return 1 instead of 3

This is on the caller to understand the output.

> maybe fixing the emoji-modifiers

Same a wcwidth it is state-less, do you really think it is worth to support state-full praising to allow modifiers?

> and adding useful instructions how to update it when a new Unicode version is released, it's OK to use in mpv IMO.

Sorry, but I won't pollute mpv codebase with instructions how to compile and run dotnet binaries, as for updating it is obvious form the output what is unused and has to be stripped and if not someone can improve the code on the update. It won't be updated every week... if ever.

> It's probably slower than it could have been for ascii chars

You are using word "slower" as if it really be of any difference. You mentioned "cache" on irc. This whole arrays fit into change, so at worse you get them on the first characters, also if you talk about ascii case, you would constantly use the same page of memory.

But sure I've added shortcut for ascii as requested. Now instead cache misses, we will get branch predictor misses...

And now it doesn't matter, it is unmeasurable difference in mpv case.

> You're welcome.

Thanks.

avih commented:
> > t's probably slower than it could have been for ascii chars
> 
> You are using word "slower" as if it really be of any difference

Did I?

Looks like you accidentally stopped reading after that sentence. Here's the complete parahraph, for your convenience.

> It's probably slower than it could have been for ascii chars, which are arguably the vast majority of the input it would have to deal with, but that's not really an issue IMO in mpv

And you already fixed that anyway. What was that comment good for? (don't answer)

About the wchar_t to uint32_t which you fixed (good):
>>  which can't hold the codepoints of all the nice Emojies from the recent decade.

> They are "broken" on Windows anyway, because there wchar_t is used ubiquitously for wide character strings. But good point, no need to impose this limitation here.

Are they broken? AFAIK mpv has no issues with codepoints above U+FFFF (16 bit) on Windows, and `wchar_t` is used correctly in Windows code in mpv. i.e. it doesn't prevent usage of non-BMP codepoints, because it correctly becomes a surrogate-pair in a wchar_t string in Windows code.

And I assume that change from wchar_t to uint32_t did fix a real issue when using this new `codepoint_width`. Do correct me if I'm wrong.

> > and adding useful instructions how to update it when a new Unicode version is released, it's OK to use in mpv IMO.
> 
> Sorry, but I won't pollute mpv codebase with instructions how to compile and run dotnet binaries, as for updating it is obvious form the output what is unused and has to be stripped and if not someone can improve the code on the update. It won't be updated every week... if ever.

I'm not the one to judge what pollution is here, but there should definitely be instructions on how to update it.

We already found one issue where your "trivial" code adjustment introduced a bug (wchar_t instead of 32 bit value), so it's better to have exact instructions so that the next person which tries to update it doesn't waste time on things which you already solved.

I think it should be obvious that code that comes from outside of mpv should include instructions on how to update it, especially when it requires some manual modifications which are not mentioned anywhere, because updates are required occasionally when Unicode has a new version, as happened with Unicode 15.1 some months ago, and Unicode 16.0 few days ago, and it will get updated again.

Also, the requirements for dotnet-sdk or visual studio is a pretty hefty IMO to update the code to a new Unicode version, especially for code which is used on all platforms in mpv and not only Windows, and especially when there are enough wcwidth implementations in highly reputable projects which don't impose any such requirements.

I won't argue this further. If others don't mind, so be it. But I do hope that other devs do mind.

kasper93 commented:
> > > t's probably slower than it could have been for ascii chars
> > 
> > 
> > You are using word "slower" as if it really be of any difference
> 
> Did I?
> 
> Looks like you accidentally stopped reading after that sentence. Here's the complete parahraph, for your convenience.
> 
> > It's probably slower than it could have been for ascii chars, which are arguably the vast majority of the input it would have to deal with, but that's not really an issue IMO in mpv
> 
> And you already fixed that anyway. What was that comment good for? (don't answer)

You have mentioned that it is 'slower' multiple times. One can only assume that it is an important fact that needs to be addressed since it is mentioned repeatedly.

> (don't answer)

I recognize that it is my issue for not being able to decipher which part of the discussion I should address. You say something, repeat it, and don't expect an answer. I'm just getting confused and lost about which areas I should focus on. Sorry about that; I will try to do better in this regard.

> About the wchar_t to uint32_t which you fixed (good):
> 
> > > which can't hold the codepoints of all the nice Emojies from the recent decade.
> 
> > They are "broken" on Windows anyway, because there wchar_t is used ubiquitously for wide character strings. But good point, no need to impose this limitation here.
> 
> Are they broken? AFAIK mpv has no issues with codepoints above U+FFFF (16 bit) on Windows, and `wchar_t` is used correctly in Windows code in mpv. i.e. it doesn't prevent usage of non-BMP codepoints, because it correctly becomes a surrogate-pair in a wchar_t string in Windows code.

I'm currently trying to address the issue regarding that. However, it's not related to this PR.

> 
> > > and adding useful instructions how to update it when a new Unicode version is released, it's OK to use in mpv IMO.
> > 
> > 
> > Sorry, but I won't pollute mpv codebase with instructions how to compile and run dotnet binaries, as for updating it is obvious form the output what is unused and has to be stripped and if not someone can improve the code on the update. It won't be updated every week... if ever.
> 
> I'm not the one to judge what pollution is here, but there should definitely be instructions on how to update it.
> 
> We already found one issue where your "trivial" code adjustment introduced a bug (wchar_t instead of 32 bit value), so it's better to have exact instructions so that the next person which tries to update it doesn't waste time on things which you already solved.
> 
> I think it should be obvious that code that comes from outside of mpv should include instructions on how to update it, especially when it requires some manual modifications which are not mentioned anywhere, because updates are required occasionally when Unicode has a new version, as happened with Unicode 15.1 some months ago, and Unicode 16.0 few days ago, and it will get updated again.

The `wchar_t` error was only mine and not related to any "porting" process. 

The problem with "instruction" is that, the upstream script may change, the exact changes may change and so on. The point is we could use C++ generated code without any modifications, but I know it wouldn't be liked so I ported it to C.

This is the entire instruction: generate the C++ code and adjust it to work with our codebase. There is no strict rule for what to change. Currently, it mostly involves removing `constexpr`. I’ve removed unused functions and added our entry point, but this doesn’t need to be repeated. As for updating, if you generate new code, replacing the tables is currently enough to make it work, though there’s no guarantee that will be true in the future. The person updating the code needs to understand it to do so effectively. Step-by-step guides are just as useful as using ChatGPT and mindlessly copying the result. Maybe it will work, maybe it won’t, but there’s no value in that until someone reviews the final outcome.

> Also, the requirements for dotnet-sdk or visual studio is a pretty hefty IMO to update the code to a new Unicode version, especially for code which is used on all platforms in mpv and not only Windows, 

Could you elaborate on this part? I would like to understand the limitations of using .NET. Which platforms are you concerned about?

> .NET is a free and open-source, managed computer software framework. It is a cross-platform successor to .NET Framework and released under the MIT License.

https://wiki.freebsd.org/.NET

.NET is widely adopted by Linux distributions, macOS, and FreeBSD, as shown above. OpenBSD may not have this support, but I don't think we have any developers actively working on OpenBSD who would be affected by this.

> and especially when there are enough wcwidth implementations in highly reputable projects which don't impose any such requirements.

It took me a significant amount of time to understand the efficient Unicode lookup algorithm and write the trie generator. While it's mentioned everywhere, there are no portable implementations available without relying on something like ICU. Then, I discovered that a generator is already maintained in a project used by millions of users, with 95k stars, so, instead of NIHing our version, I decided to use the existing generator.

> there are enough wcwidth implementations

Agree to disagree on this point. Many of them are outdated and not maintained. The main point is that by actually using Grapheme Clusters, we can support it properly. Granted, I skipped the state part because I don't think it's useful, but wcwidth variants are inherently limited by this anyway. I think I will implement full resolver, to remove this ambiguity.

avih commented:
> You have mentioned that it is 'slower' multiple times. One can only assume that it is an important fact that needs to be addressed since it is mentioned repeatedly.

It was mentioned exactly twice. Once on IRC, and once here, and in both cases with an explicit comment that it doesn't matter in mpv. Here's what I said on IRC:

> avih: (but this was more of an anecdote, in mpv it's far from hot code, and no one would notice even if it was 100x slower)

So you really can't say "You are using word "slower" as if it really be of any difference", because on both places I mentioned it I explicitly also said it will NOT make any difference.

But it's also worth mentioning anyway that it's slower, because it is. Now, you can change that, or not. It doesn't matter, but I mentioned it so that you become aware of this difference with other implementations, and so that you can decide for yourself about it. And you decided to change that. Good for you.

Just don't say that I think it matters a lot, because I said exactly the opposite.

You're trying to generate confrontations, and not for the first time. Please don't do that.

> It took me a significant amount of time to understand the efficient Unicode lookup algorithm and write the trie generator. While it's mentioned everywhere, there are no portable implementations available without relying on something like ICU. Then, I discovered that a generator is already maintained in a project used by millions of users, with 95k stars, so, instead of NIHing our version, I decided to use the existing generator.

That's fair, but I didn't check how much code the trie compressor has. I don't think it should be too much, but I don't know that for a fact. If it's not too much (say, 50-100 LOC in C), and assuming we can't find an existing implementation in more easily-portable language, then I think it wouldn't be too bad to host it here.

But I can't say much about the compressor because I've not looked at existing implementations.

> The main point is that by actually using Grapheme Clusters, we can support it properly

Not sure what that means exactly (yes, I do know what Grapheme clusters are, just not what that statement means), but if this really is the "main point" which is needed to make `codepoint_width` retrurn the values which it currently returns, then rejoice!, because the data which `codepoint_width` returns is trivial to extract from the unicode data files.

Here are 3 lines of shell script which extract from the unicode data files exactly the data which `codepoint_width` returns:
```sh
# print wide and ambiguous ranges
sed -e '/^#/d' -e 's/[;#]/ /g' < ./EastAsianWidth.txt |
awk '$2 ~ /[FW]/ { print $1, 2 }
     $2 ~ /A/    { print $1, 3 }'

# print zero-width codepoints
awk -F";" '$3 ~ /Me|Mn|Cf/ { print $1, 0 }' < ./UnicodeData.txt
```

And here's 40 lines shell script which can run anywhere, even on your router, or with busybox, which processes the Unicode source files and produces a `int simple_codepoint_width(uint32_t)` implementation which returns identical values like this PR's `codepoint_width` for every one of the 0x110000 codepoints:
<details>
<summary>codepoint-width-gen.sh</summary>

```sh
#!/bin/sh

# Download the latest Unicode data files:
#   https://www.unicode.org/Public/UCD/latest/ucd/UnicodeData.txt
#   https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt

print_non_width1_codepoints() { 
    # print wide and ambiguous ranges
    sed -e '/^#/d' -e 's/[;#]/ /g' < ./EastAsianWidth.txt |
    awk '$2 ~ /[FW]/ { print $1, 2 }
         $2 ~ /A/    { print $1, 3 }'

    # print zero-width codepoints
    awk -F";" '$3 ~ /Me|Mn|Cf/ { print $1, 0 }' < ./UnicodeData.txt
}

as_decimal_ranges() {
    while read range width; do
        echo $((0x${range%%.*})) $((0x${range##*.})) $width
    done
}

print_non_width1_codepoints |
as_decimal_ranges |
awk '
    { for (i = $1; i <= $2; i++) w[i] = $3 + 1 }  # store width + 1

    END {
        # manually override box drawing range 0x2500..0x259F to normal-width
        # https://github.com/microsoft/terminal/blob/a7e47b71/src/tools/GraphemeTableGen/Program.cs#L305
        for (i = 9472; i <= 9631; i++) w[i] = 1 + 1

        indent = "\n        "

        print  "int simple_codepoint_width(uint32_t ucs)"
        print  "{"
        printf "    static const char widths[] = {"

            for (i = 0; i < 1114112; i++)  # 0x110000
                printf "%s%d,", (i % 32 ? "" : indent), (w[i] ? w[i] - 1 : 1)

        print  ""
        print  "    };"
        print  "    return widths[ucs];"
        print  "}"
    }
'
```

</details>

So if we have an easily-portable trie-compressor we can use it to compress the array and end up with O(1) lookup solution.

Or with about 5 more lines at the awk script, we can print the data as ranges of wide/ambiguous/zero codepoints, and use binary search like many other implementations do, and then there are no external dependencies at all.

kasper93 commented:
> You're trying to generate confrontations, and not for the first time. Please don't do that.

I will try my best to improve, thank you for you feedback.

> And here's 40 lines shell script which can run anywhere, even on your router

I believe it’s unproductive to continue in this manner.

Let’s focus on the topic at hand. I won’t be generating anything on my router or detecting Unicode codepoint width, especially considering I don’t believe it has 1MB of RAM.

Regarding #14466, it wasn’t merged because you expressed your rejection on IRC and in your response. I’ve now developed a better, industry standard, solution that uses 10KB instead of the currently proposed 1MB, resulting in a 99% size reduction.

I feel like there’s a consistent effort to discredit this work. If merging isn’t an option, that’s okay. We can move forward from here.

> Or with about 5 more lines at the awk script, we can print the data as ranges of wide/ambiguous/zero codepoints, and use binary search like many other implementations do

Yes, they do, and this is a naive approach. I see no reason to use a naive method when there is an industry standard solution for Unicode codepoint lookup. For more details and related work, please see https://unicode-org.github.io/icu/design/struct/utrie.

> And better yet, we can remove the ambigous codepoints data because the code converts them to width or 1 anyway. And we can return them once we need ambiguous width. Because it's easy when everything you need to update or modify the implementation fits in one page of code which can run anywhere, even on your phone.

Please focus on this PR and the proposed changes, specifically the problems and limitations that need to be addressed. I understand that you may have different ideas, but I haven’t seen a PR from you yet. I’d appreciate your guidance on how I can improve this PR to gain your approval, without going into comments that may be perceived as dismissive, like generating it on the router.

kasper93 commented:
I've updated the code with full grapheme clusters support and added some tests.

avih commented:
> I will try my best to improve, thank you for you feedback.

Great.

> I’ve now developed a better, industry standard, solution that uses 10KB instead of the currently proposed 1MB, resulting in a 99% size reduction.

Oops. But you couldn't hold yourself back, could you?

I explicitly said:

> (obviously simple_codepoint_width is also O(1), but it does use 1M ram, which is arguably more than we want, so on its own it's not good enough).

So you know perfectly well I never proposed that. It was an example of how trivial it is to generate the data which `codepoint_width` returns, and to generate a `simple_codepoint_width` function implementation which returns identical values as `codepoint_width` in this PR.

But still try to hit a punch. Good for you.

What I did propose, however, was to use binary search which is much simpler to generate, because we both agree (I think) that terminal string width performance is not a factor in the mpv context (on that level of the diff between binary search and trie lookups, not in general) .

If we did binary search, then that would be 2K data, which is 4x smaller than your suggestion. I don't think it matters (keep that in mind please), but since you were kind to mention it, allow me to pay back the kindness.

> I believe it’s unproductive to continue in this manner.

At least we agree on that. I'm out of this PR.

kasper93 commented:
> So you know perfectly well I never proposed that.

I understand what you’re saying, but it seems like there are many unrelated points being made. This makes it a bit difficult to understand what you need or want.

What was the purpose of the router remark? What was the point of posting code to generate an "unusable" 1MB lookup table? I feel like you are derailing the discussion by venturing into unrelated areas.

> At least we agree on that. I'm out of this PR.

Classic.

avih commented:
> This makes it a bit difficult to understand what you need or want.

One thing. To make it easy to update to a new Unicode version once it comes out.

If I get that right, you consider mono/dotnet-sdk/visual-studio/whatever needed to run the update "easy", and also think it's easy to perform some manual modifications of the resulting code. I don't. We disagree on what "easy" is.

And I gave examples of things I think would be much easier. On the Matter of obtaining the exact data which `codepoint_width` uses, and on generating a function which returns identical values, with an updater that is portable enough so that it can run almost anywhere out of the box (that was the reference to phone/router).

I said I prefer to have 3rd party trie-compressor which uses more portable language, or maybe in-house if you feel like it (because you said you did that already but don't want NIH, but I think it's OK if it's small enough).

I also said that binary search would also be fine if we don't have a trie-compressor, for the reasons I mentioned.

That's it.

Easy to update, and keep in mind that we don't need uber-performance in our context, and that "very good" performance is likely more than good enough. The rest is derived from that.

kasper93 commented:
> > This makes it a bit difficult to understand what you need or want.
> 
> One thing. To make it easy to update to a new Unicode version once it comes out.
> 
> If I get that right, you consider mono/dotnet-sdk/visual-studio/whatever needed to run the update "easy", and also think it's easy to perform some manual modifications of the resulting code. I don't. We disagree on what "easy" is.

Fair. In this case could you specify what are the requirements? For example would it be "easy" if we had a generator in Rust, Go, Python or JS (through node)? Or does it have bash script? And reasoning why.

To offer a solution, I need to fully grasp the problem first. From my perspective, using dotnet-sdk doesn't present any significant issue, given that it's available on all major Linux distributions, macOS, FreeBSD, and Windows. In fact, I find the process smoother than it might be with Rust, for example. It’s comparable to setting up any standard development toolchain with package manager of your distro.

Just to be clear. Here is all the steps required to generate the source:

``` bash
sudo apt-get install dotnet-sdk-8.0
dotnet build GraphemeTableGen.csproj
./bin/Debug/net8.0/GraphemeTableGen
```

And here are the manual adjustment that I mad to the generated code. Which are limited to adding licence header of generated code and removing C++ keywords like `constexpr` and `auto`.
``` diff
diff --git a/misc/codepoint_width.c b/misc/codepoint_width.c
index abe9dd1136..981b55159a 100644
--- a/misc/codepoint_width.c
+++ b/misc/codepoint_width.c
@@ -9,10 +9,15 @@
 typedef uint_least32_t char32_t;
 #endif
 
-// Generated by GraphemeTableGen
+// Generated by GraphemeTableGen (https://github.com/microsoft/terminal/blob/a7e47b711a2adc7b9e80eddea8168089f7d3b11e/src/tools/GraphemeTableGen/Program.cs)
 // on 2024-09-21T14:12:16Z, from Unicode 16.0.0, 9285 bytes
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Manually adjusted for C compatiblity
+
 // clang-format off
-static constexpr uint16_t s_stage0[] = {
+static const uint16_t s_stage0[] = {
     0x0000, 0x0020, 0x0040, 0x0060, 0x0080, 0x009f, 0x00bf, 0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00ca,
     0x00ca, 0x00ca, 0x00ca, 0x00ca, 0x00d8, 0x00f8, 0x010a, 0x010e, 0x010b, 0x0108, 0x0113, 0x0133, 0x0153, 0x0153, 0x0153, 0x016f,
     0x018f, 0x01a7, 0x01c7, 0x01e7, 0x0133, 0x0133, 0x0205, 0x0133, 0x0133, 0x0133, 0x0133, 0x0133, 0x0221, 0x0236, 0x00ca, 0x00ca,
@@ -48,7 +53,7 @@ static constexpr uint16_t s_stage0[] = {
     0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153,
     0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x0153, 0x03e7,
 };
-static constexpr uint16_t s_stage1[] = {
+static const uint16_t s_stage1[] = {
     0x0000, 0x0004, 0x000c, 0x0014, 0x001c, 0x0024, 0x002a, 0x0031, 0x002a, 0x0037, 0x002a, 0x003f, 0x0047, 0x0049, 0x004f, 0x0057, 0x005f, 0x0065, 0x006d, 0x002a, 0x002a, 0x002a, 0x0073, 0x007b, 0x0083, 0x008a, 0x002a, 0x0091, 0x0098, 0x009f, 0x00a3, 0x00aa,
     0x00b2, 0x00b8, 0x00be, 0x00c5, 0x00cd, 0x00d5, 0x00dd, 0x00e5, 0x00ed, 0x00f5, 0x00fd, 0x0105, 0x010d, 0x0115, 0x011d, 0x0125, 0x012d, 0x0135, 0x013d, 0x0145, 0x014d, 0x0155, 0x015d, 0x0164, 0x016b, 0x0173, 0x0175, 0x017d, 0x0182, 0x018a, 0x0192, 0x019a,
     0x019d, 0x01a5, 0x01ad, 0x002a, 0x01b5, 0x01b9, 0x01bd, 0x01c2, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x01ca, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x002a, 0x01d0, 0x01d7, 0x01de, 0x01e6,
@@ -83,7 +88,7 @@ static constexpr uint16_t s_stage1[] = {
     0x076e, 0x076e, 0x076e, 0x076e, 0x076e, 0x076e, 0x076e, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc,
     0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x02bc, 0x077e,
 };
-static constexpr uint16_t s_stage2[] = {
+static const uint16_t s_stage2[] = {
     0x0000, 0x0000, 0x0000, 0x0000, 0x0008, 0x0008, 0x0008, 0x0008,
     0x0008, 0x0008, 0x0008, 0x0009, 0x0000, 0x0000, 0x0000, 0x0000,
     0x0011, 0x0018, 0x0020, 0x0022, 0x002a, 0x0008, 0x0029, 0x0030,
@@ -326,7 +331,7 @@ static constexpr uint16_t s_stage2[] = {
     0x0089, 0x0089, 0x0089, 0x0089, 0x0000, 0x0000, 0x0092, 0x0092,
     0x0092, 0x0092, 0x0092, 0x0092, 0x0092, 0x0863,
 };
-static constexpr uint8_t s_stage3[] = {
+static const uint8_t s_stage3[] = {
     0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41,
     0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40,
     0x41, 0x40, 0xc0, 0x40, 0x40, 0xc0, 0x40, 0x40,
@@ -598,7 +603,7 @@ static constexpr uint8_t s_stage3[] = {
     0x41, 0x41, 0x41, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0,
     0xc0, 0x40, 0x40,
 };
-static constexpr uint32_t s_joinRules[2][16] = {
+static const uint32_t s_joinRules[2][16] = {
     {
         0b00000011110011111111111111001111,
         0b00001111111111111111111111111111,
@@ -636,25 +641,25 @@ static constexpr uint32_t s_joinRules[2][16] = {
         0b00000000000000000000000000000000,
     },
 };
-constexpr int ucdLookup(const char32_t cp) noexcept
+static int ucdLookup(const char32_t cp)
 {
-    const auto s0 = s_stage0[cp >> 11];
-    const auto s1 = s_stage1[s0 + ((cp >> 6) & 31)];
-    const auto s2 = s_stage2[s1 + ((cp >> 3) & 7)];
-    const auto s3 = s_stage3[s2 + ((cp >> 0) & 7)];
+    const uint16_t s0 = s_stage0[cp >> 11];
+    const uint16_t s1 = s_stage1[s0 + ((cp >> 6) & 31)];
+    const uint16_t s2 = s_stage2[s1 + ((cp >> 3) & 7)];
+    const uint8_t s3 = s_stage3[s2 + ((cp >> 0) & 7)];
     return s3;
 }
-constexpr int ucdGraphemeJoins(const int state, const int lead, const int trail) noexcept
+static int ucdGraphemeJoins(const int state, const int lead, const int trail)
 {
-    const auto l = lead & 15;
-    const auto t = trail & 15;
+    const int l = lead & 15;
+    const int t = trail & 15;
     return (s_joinRules[state][l] >> (t * 2)) & 3;
 }
-constexpr bool ucdGraphemeDone(const int state) noexcept
+static bool ucdGraphemeDone(const int state)
 {
     return state == 3;
 }
-constexpr int ucdToCharacterWidth(const int val) noexcept
+static int ucdToCharacterWidth(const int val)
 {
     return val >> 6;
 }
```

I don't see any of those steps hard and when updating the generated code, it would have been easy to see the difference after copy of the code.

> 
> And I gave examples of things I think would be much easier. On the Matter of obtaining the exact data which `codepoint_width` uses, and on generating a function which returns identical values, with an updater that is trivial to operate, small enough to fit in that coment, and portable enough so that it can run almost anywhere out of the box (that was the reference to phone/router).

To be honest, I don't see how using bash with the help of sed and awk makes it easier to parse an XML file. Not to mention generating tries and join tables to support clusters. Either way, I've already asked about this above.

The point is, I don't see the benefit of maintaining a custom parser/generator in the mpv tree instead of using a well-established one and leveraging the power of open source. We can use the work of others instead of reinventing the wheel. It would certainly be not easier to maintain own script.

That said, I'm open to suggestions. I'm not insisting on using GraphemeTableGen. If you can provide and maintain a script that achieves the same result, passes the test/codepoint_width.c, and does it better, I see no reason not to use that. At the same time, I still don't understand why using GraphemeTableGen is considered "bad."

Also, to circle back to the "slow" topic, I don't think mpv is significantly affected by minor performance differences. However, something 100x slower, as you mentioned, would be, in my opinion, unacceptable. Even currently, mpv can slow down when a lot of logs are printed, so I don't want to add more CPU overhead.

avih commented:
> For example would it be "easy" if we had a generator in Rust, Go, Python or JS (through node)?

For me, all of those would be reasonable enough. I still think think C or `sh` (not bash) are more portable, but your suggestions don't sound bad to me. As long as it's similarly easy to develop the tool, the least an mpv developer needs to install additionally to what they already have for mpv - the better IMO.

So of those options, probably python because mpv devs have that already for meson.

> Just to be clear. Here is all the steps required to generate the source:

That's great to have (no sacrasm).

> And here are the manual adjustment that I mad to the generated code. Which are limited to adding licence header of generated code and removing C++ keywords like `constexpr` and `auto`.

This is also great to have, and probably sufficient for the next person to have a good idea how they need to modify the output.

Next question is, where do we put this diff so that it can be found when the time comes for the next update? (commit message? some file at the repo? etc).

> I don't see any of those steps hard and when updating the generated code,

Agreed, especially when they're listed clearly, as you did just now, and people don't have to guess.

> The point is, I don't see the benefit of maintaining a custom parser/generator in the mpv tree instead of using a well-established one and leveraging the power of open source. We can use the work of others instead of reinventing the wheel. It would certainly be not easier to maintain own script.

Agreed largely, though there are many forms to reuse 3rd party projects. One of them is simply to take the pre-generated file and edit it (similar to the diff you posted), with instructions if someone really wants to re-generate it themselves. This is much easier IMO, without losing much (except maybe waiting a bit till the 3rd party project re-generates their files).

>  I'm not insisting on using GraphemeTableGen

Not sure what this refers to exactly. Is that the usage of that Program.cs generator? or only the additional Grapheme cluster code which is getting generated and you added in one of the force-pushes? I actually like it, so nothing aginst it here (though AFAIK most projects which use `wcwidth` don't also use something at a higher level like the grapheme cluster tables/code, but it's nice). Probably worth mentioning here the suckless libgrapheme.

Anyway, I don't really mind which code gets used, as long as it's reasonable, and our requirements are quite modest anyway. As I said, my main concern is that it would be easy to update when the time comes. For the rest, including the specific source to use and its quality, I trust your judgment.

> Also, to circle back to the "slow" topic, I don't think mpv is significantly affected by minor performance differences. However, something 100x slower, as you mentioned, would be, in my opinion, unacceptable. Even currently, mpv can slow down when a lot of logs are printed, so I don't want to add more CPU overhead.

Yeah, maybe 100x slower was a bit of an overshoot, but it does sound nice!

But anyway, keep in mind that the vast majority of mpv terminal log output is ascii, and this would have the same performance in all implementations, and it will be very very quick.

But if we can easily generate compressed tries, I'm all for it. It's cool.

Even with the binary search (tested my own "smart" search - the code in #14466) vs tries-lookup, even if all the input is non-ascii, which does require the search, it's definitely in the same ballpark - not even twice slower (but it is measurably slower), and both are very very fast, way more than mpv needs IMO (but obviously it's not the only bottleneck and it does add up, so if it's easy to get uber-performance, it's preferable).

With these instructions which you posted above, especially if there's also a way to take a pregenerated file from someplace and just edit it (which is explained someplace), then I don't think it's too bad.

The return values of codepoint_width are very straight forward as far as Unicode tables go, and I trust the tries are generated correctly, and the grapheme clusters thing is also nice to have, so the code LGTM (but do check that emoji combo which I posted - that it returns 2 and not 4).

Just find a useful way to make these detailed update instructions accessible, and this is not a bad solution IMO.

avih commented:
> Even currently, mpv can slow down when a lot of logs are printed, so I don't want to add more CPU overhead.

Wait, how is log output related to graphem clusters width calculations? surely the parts which print in volumes large enough to slow down mpv don't need to care about terminal width of strings?!

kasper93 commented:
> That's great to have (no sacrasm).

Already has been posted few days ago https://github.com/mpv-player/mpv/pull/14894#issuecomment-2365244990

> Next question is, where do we put this diff so that it can be found when the time comes for the next update? (commit message? some file at the repo? etc).

Nowhere, because as you may know diff will not apply after the target code changes. Commit message already mention the required changes.

> Not sure what this refers to exactly. Is that the usage of that Program.cs generator?

Yes, I refer to [GraphemeTableGen](https://github.com/microsoft/terminal/tree/main/src/tools/GraphemeTableGen), Program.cs is just main source code, the project is named `GraphemeTableGen`.

kasper93 commented:
> (but do check that emoji combo which I posted - that it returns 2 and not 4).

Could you remind me which one? I've just tested one of your examples and it passes the test.
```
assert_int_equal(W("\U0001F469\U0001F3FB"), 2);
```
In fact there is already similar case tested.

> Wait, how is log output related to graphem clusters width calculations? surely the parts which print in volumes large enough to slow down mpv don't need to care about terminal width of strings?!

I don't understand what do you mean. But whatever you are trying to say here is not related to this PR in anyway.

avih commented:
>>> Even currently, mpv can slow down when a lot of logs are printed, so I don't want to add more CPU overhead.

>> Wait, how is log output related to graphem clusters width calculations? surely the parts which print in volumes large enough to slow down mpv don't need to care about terminal width of strings?!

> I don't understand what do you mean. But whatever you are trying to say here is not related to this PR in anyway.

You said that a lot of log printing already can slow down mpv, so I assumed you considered it relevant, but I don't understand how.

So how is that log-print-cpu comment relevant to this PR?

avih commented:
> (Side note: existing table generators exist, like https://raw.githubusercontent.com/GNOME/glib/master/glib/gen-unicode-tables.pl. PuTTY has its own implementation of `wcwidth` which uses that.)

Yeah. There are many of those around. In case you missed it, I posted above a 1 page script which generates a `codepoint_width()` function directly from the Unicode data files, and which returns identical values as the function in this PR - here https://github.com/mpv-player/mpv/pull/14894#issuecomment-2366596192

kasper93 commented on misc/codepoint_width.c:
```diff
@@ -0,0 +1,758 @@
+// Licensed under the MIT license.
+
+// Manually adjusted for C compatiblity
+
```
Added

kasper93 commented on misc/codepoint_width.c:
```diff
@@ -0,0 +1,758 @@
+        0b00000000000000000000000000000000,
+    },
+};
+static int ucdLookup(const char32_t cp)
```
I want to minimize unneeded changes in generated code. We can modernize our `bstr_decode_utf8` later, but it is outside of the scope of this PR.

One other improvement for `bstr_decode_utf8` would be to not abort processing or invalid chars and just skip them over. Currently our processing will just stop once they are found. Still not the scope of this PR.

kasper93 commented on misc/codepoint_width.c:
```diff
@@ -0,0 +1,758 @@
+};
+static const uint32_t s_joinRules[2][16] = {
+    {
+        0b00000011110011111111111111001111,
```
mpv is not exactly free from extensions and this works fine on all compilers that we support.

kasper93 replied:
I think it the case above it actually were throwing warnings and failing to compile. Not saying using non standard extensions is good, but we need to be pragmatic here. I really want to avoid vendoring generator / rewriting / maintaining it. As you may have noticed. Let say it another way, I won't do that myself, I won't oppose if anyone else wants to do it though.

Changing this from binary representation is small change, but once we start doing it the process of updating is no longer "easy", because require modification of generation code and so on.

kasper93 replied:
> Right now the generation process requires some manual fiddling anyway. I think it's a good idea to add a script in TOOLS which automates the whole process providing that .NET is installed. Once that's on the table the replacement can be done in the script.

I think we're focusing too much on automating something that's only updated every 1-2 years at best. While it would be ideal to fully automate and just watch the code being written, it's not that simple.

Maybe my reasoning isn't clear, so let me explain. I'm not interested in maintaining this generator in the mpv tree because it’s not necessary. I understand you're not suggesting that exactly, but having a script that does generation and adjustment automatically, is not easy either. Sure, s/constexpr/const/ is easy, but replacing auto is not. We'd need to build an AST to figure out the correct type for auto (or ask an LLM). It’s not worth the effort. So, we're back to either creating our own generator or modifying the upstream one, which I don't think is needed.

The issue is that in 1-2 years, when this script is used again, the upstream version will likely be updated, rendering any static patches we create now useless. Since updates are so infrequent, we can afford to revisit this topic in a couple of years and update the version and review the changes that may have happened upstream.

I'm happy that there's an easy way to generate the code. It might not be a perfect drop-in replacement, but the "manual fiddling" only involves changing a dozen of lines. We can't predict what will happen in the future... there could be a completely different algorithm needed.

It's not that I don't want to automate this. I do I like automation, as you probably know if you've followed my CI work. I just don't know how to do it without maintaining our own script, which I’d rather avoid for something used so rarely.

That said, any suggestions or code examples on how to achieve this are very welcome.

But if this is a blocker for the PR, so be it. It wouldn’t be the first time a nice feature is blocked over a mundane issue.

> For now it's OK, but it will become a problem if we bump the warning level to 3 which enables -Wpedantic.

Don't raise the bar for this PR. It doesn't matter at this point. There are still unresolved warnings at the current level that are being suppressed, and using pedantic would generate even more. I'm all for resolving warnings, but if it impacts other aspects, it's not always beneficial to force fixing them.

avih replied:
> I think we're focusing too much on automating something that's only updated every 1-2 years at best

The last two updates were few months apart (15.1 and 16.0).

kasper93 replied:
> The last two updates were few months apart (15.1 and 16.0).

12 months apart to be exact.

kasper93 replied:
> If it's the case that the frequency of the updates doesn't justify the efforts for automation (which is my initial thought), then what's the issue with manually converting the numeric format for these infrequent situations? Compared to potential incompatible upstream changes that require more manual fix, this effort seems minuscule.

Fair point. I was focused on minimizing changes as much as possible, but I can agree to do this change too. Changed now to base 16.

avih replied:
> Downgrading the algorithm to bisection without clusters seems unnecessary.

I think you miss that the Grapheme clusters code is mostly no-op. wcwidth alone can handle Grapheme clusters very well, with very few exceptions.

These exceptions are very few codepoints which do have a width of their own (e.g. 2), but their main usage is combined with others as modifiers, typically/exclussively with emojis, at which case their width (in a series following an emoji) becomes 0.

If the wcwidth implementation assigns them a width of 0 (which for instance the code at #14466 does), then this covers what Grapheme clusters analysis does. It does leave out that it will not detect of these typically-modifiers on its own, but that's a very very rare use case in practice.

For instance emoji skin tone modifiers on their own would look like a sqare in a specific skin tone, but that doesn't really happen with normal content, because mostly it modifies an emoji, so if wcwidth returns 0 for these modifiers, then you don't need the grapheme clusters analysis for these cases.

Not completely perfect, but effectively perfect.

And the added value is that you need a one page script to generate it instead of dot-net sdk, and can tweak it however you see fit, because it's a tiny generator.

The balance is between cost and value. And when the value is effectively the same, it comes down to cost.

kasper93 replied:
> And when the value is effectively the same...

I don't know if you're trying to troll me or if something is genuinely wrong here. I've explained multiple times why the value is not the same. Grapheme clusters are not supported, and the bisection method is naive and not optimal for this task. You can't ignore all the facts in the trade-off analysis and claim that solution A is the same as solution B just because they work the same "90%" of the time.

You are spending whole paragraphs of text explaining why inferior solution is "almost the same", "not completely perfect", what is the reason for that? Why current solution cannot be used?

> The balance is between cost and value.

Exactly. The cost of updating this code every 1-2 years is almost zero. That's why I don't see why using inferior solution would be preferable? Maintaining even simple script would have higher cost than just running existing solution. I still didn't see any argument why using this generator is "costly" or "hard" except initial misunderstanding about .NET.

avih replied:
Nevermind. Please ignore my comment. I'm leaving it here only for posterity.

kasper93 replied:
As I mentioned earlier, I'm open to other solutions. If you submit a PR with code that is better by your standards, I'd be happy to replace the current implementation. However, at this point, I don't see any major issues with the current approach that would prevent it from being used.

kasper93 commented on misc/codepoint_width.c:
```diff
@@ -0,0 +1,758 @@
+
+int term_disp_width(bstr str)
+{
+    static const int ambiguous_width = 1;
```
This is expected. Do you have examples where this matters? If you read next line, 1 is more accurate fallback for ambiguous width. We use 1 width by default, but once we see `0xFE0F` to convert those unqualified Emojis we promote them to 2 width. That said I think it works as expected and best we can do. If you have cases where it breaks, please share I will adjust accordingly.

kasper93 replied:
Are we still using those circles? Didn't we want to remove them?

Hmm, one small thing is that msg.c will actually work better when we underestimate the width, instead of overestimating. Basically in our case current way will work better in practice, equally bad, but "better" in the sense it will not eat lines and instead duplicate.

kasper93 replied:
> I didn't have a strong opinion on the usage of circles instead of ASCII symbols, as long as the same width is obtained for active and inactive states.

No, guido explained me the situation on IRC. I recalled topic wrong.

> If overestimating causes lines being eaten which can't be easily solved, then this is OK for now, but I would consider making it an option later.

I will keep it as is if you don't mind, this PR is very hot already, I don't want to add any more changes here. We can improve later, if we manage to merge this guy first. Hope you understand.

kasper93 commented:
@na-na-hi: Do you have any more concerns about the code in this PR?

kasper93 commented:
This is the best I'm able to do. I know not everyone is happy, but we need to move forward somehow. I'm always open for improvements to current version.
